x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/profile/SuggestProfilesRequest.java

```diff
@@ -66,6 +82,95 @@ public ActionRequestValidationException validate() {
         if (size < 0) {
             validationException = addValidationError("[size] parameter cannot be negative but was [" + size + "]", validationException);
         }
+        if (hint != null) {
+            validationException = validateHint(validationException);
+        }
+        return validationException;
+    }
+
+    private ActionRequestValidationException validateHint(ActionRequestValidationException validationException) {
+        final List<String> uids = hint.getUids();
+        final Map<String, List<String>> labels = hint.getLabels();
+
+        if (uids == null && labels == null) {
+            return addValidationError("[hint] parameter cannot be empty", validationException);
+        }
+
+        if (uids != null && uids.isEmpty()) {
+            validationException = addValidationError("[uids] hint cannot be empty", validationException);
+        }
+
+        if (labels != null) {
+            if (labels.size() != 1) {
+                return addValidationError(
+                    "[labels] hint supports a single key, got [" + Strings.collectionToCommaDelimitedString(labels.keySet()) + "]",
+                    validationException
+                );
+            }
+
+            final List<String> values = labels.values().iterator().next();
+            if (values.isEmpty()) {
+                return addValidationError("[labels] hint cannot be empty", validationException);
+            }
+        }
         return validationException;
     }
+
+    public static class Hint implements Writeable {
+        @Nullable
+        private final List<String> uids;
+        @Nullable
+        private final Map<String, List<String>> labels;
+
+        @SuppressWarnings("rawtypes")
+        public Hint(List<String> uids, Map<String, Object> labelsInput) {
+            this.uids = uids == null ? null : List.copyOf(uids);
+            // Convert the data type. Business logic, e.g. single key, is enforced in request validation
+            if (labelsInput != null) {
+                final HashMap<String, List<String>> labels = new HashMap<>();
+                for (Map.Entry<String, Object> entry : labelsInput.entrySet()) {
+                    final Object value = entry.getValue();
```

Comment:
Very optional but I think following structure is a little easier to parse:

```java
if (value instanceof final String stringValue) {
    labels.put(entry.getKey(), List.of(stringValue));
} else if (value instanceof final List listValue) {
    final ArrayList<String> values = new ArrayList<>();
    for (Object v : listValue) {
        if (v instanceof final String stringValue) {
            values.add(stringValue);
        } else {
            throw new IllegalArgumentException("[labels] hint supports either string value or list of strings");
        }
    }
    labels.put(entry.getKey(), List.copyOf(values));
} else {
    throw new IllegalArgumentException("[labels] hint supports either string or list of strings as its value");
}
```

Reply:
👍